File: part1.py
import numpy as np
import scipy.stats as stats

# Parameters
n_simulations = 10000   # Number of simulations
claim_count_distribution = "negative_binomial"  # Non-Poisson distribution for claim counts
claim_cost_distribution_1 = "exponential"  # Satisfying Cramér's condition
claim_cost_distribution_2 = "lognormal"  # Not satisfying Cramér's condition

# Negative Binomial Distribution Parameters
r = 10  # Number of successes
p = 0.5  # Probability of success

# Exponential Distribution Parameter
lambda_exp = 0.3  # Rate parameter

# Lognormal Distribution Parameters
mu = 0  # Mean
sigma = 1  # Standard deviation

# Monte Carlo Simulation
total_costs_1 = []
total_costs_2 = []
for _ in range(n_simulations):
    # Generate number of claims
    n_claims = np.random.negative_binomial(r, p)

    # Generate claim costs
    costs_1 = np.random.exponential(scale=1/lambda_exp, size=n_claims)
    costs_2 = np.random.lognormal(mean=mu, sigma=sigma, size=n_claims)

    # Sum costs to get total cost
    total_costs_1.append(np.sum(costs_1))
    total_costs_2.append(np.sum(costs_2))

# Calculate extreme quantiles
quantile_95_1 = np.percentile(total_costs_1, 95)
quantile_99_1 = np.percentile(total_costs_1, 99)
quantile_95_2 = np.percentile(total_costs_2, 95)
quantile_99_2 = np.percentile(total_costs_2, 99)

print(quantile_95_1)

print(quantile_99_1)

print(quantile_95_2)

print(quantile_99_2)

File: part1.R
library(MASS) # pour la distribution log-normale
library(stats) # pour les fonctions de distribution

# Paramètres
n_simulations <- 10000
r <- 10  # Nombre de succès pour la distribution binomiale négative
p <- 0.5  # Probabilité de succès
lambda_exp <- 0.3  # Paramètre pour la distribution exponentielle
mu <- 0  # Moyenne pour la distribution log-normale
sigma <- 1  # Écart-type pour la distribution log-normale

# Simulation Monte Carlo
simulate_costs <- function(distribution, n_simulations, r, p, lambda_exp, mu, sigma) {
  total_costs <- numeric(n_simulations)
  for (i in 1:n_simulations) {
    n_claims <- rnbinom(1, size = r, prob = p)
    if (distribution == "exponential") {
      claim_costs <- rexp(n_claims, rate = lambda_exp)
    } else if (distribution == "lognormal") {
      claim_costs <- rlnorm(n_claims, meanlog = mu, sdlog = sigma)
    }
    total_costs[i] <- sum(claim_costs)
  }
  return(total_costs)
}

total_costs_1 <- simulate_costs("exponential", n_simulations, r, p, lambda_exp, mu, sigma)
total_costs_2 <- simulate_costs("lognormal", n_simulations, r, p, lambda_exp, mu, sigma)

# Calcul des quantiles extrêmes
quantile_95_1 <- quantile(total_costs_1, 0.95)
quantile_99_1 <- quantile(total_costs_1, 0.99)
quantile_95_2 <- quantile(total_costs_2, 0.95)
quantile_99_2 <- quantile(total_costs_2, 0.99)

print(c(quantile_95_1, quantile_99_1, quantile_95_2, quantile_99_2))
